File: site/tokenizer/check_test_vectors.mjs
/**
 * §16.6's gate, the JavaScript half. It makes the same assertions as
 * `check_test_vectors.py` — deliberately, because the point of the gate is that both
 * implementations reach the same verdict about the same file, and a checker that
 * checked something else on this side would hide exactly the drift it exists to find.
 *
 *   node check_test_vectors.mjs [directory]
 *
 * Node is used only for reading files and exiting; the tokenizer itself imports
 * nothing and runs unchanged in a Worker.
 */
import { createHash } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { Tokenizer } from './isabelle_tokenizer.js';

const HERE = dirname(fileURLToPath(import.meta.url));

// The features §16.5 requires by name. A vector file missing any of them is not a
// gate, whatever its size.
const REQUIRED_FEATURES = [
  '16.2', '5.3', 'nfd', 'u007f', 'subsup_without_a_fold_entry', 'separator_only',
  'boundary_character', 'fallback_clause', 'private_use_escape',
  'escape_against_ascii_symbolic', 'astral_symbol', 'adjacent_fold_markers',
  'escape_scanning', 'numeric_class', 'empty', 'real_expression', 'real_name',
];

// U+0085, U+2028 and U+2029 terminate lines for Python's str.splitlines and for a
// good many other readers, and real Isabelle text contains them; the generator
// escapes them, and this asserts it did.
const STRAY_LINE_BREAK = /[\u0085\u2028\u2029\r]/;

const sha256 = (buffer) => createHash('sha256').update(buffer).digest('hex');

const parseHistory = (text) => text.split('\n')
  .map((line) => line.trim())
  .filter((line) => line && !line.startsWith('#'))
  .map((line) => {
    const fields = Object.fromEntries(line.split(/\s+/)
      .filter((part) => part.includes('='))
      .map((part) => [part.slice(0, part.indexOf('=')), part.slice(part.indexOf('=') + 1)]));
    return {
      count: Number(fields.count),
      sha256: fields.sha256,
      ruleChange: line.includes('rule-change:'),
      line,
    };
  });

const sameArray = (a, b) => a.length === b.length && a.every((x, i) => x === b[i]);

export function main(directory = HERE) {
  const problems = [];
  const meta = JSON.parse(readFileSync(join(directory, 'test_vectors.meta.json'), 'utf8'));
  const body = readFileSync(join(directory, 'test_vectors.jsonl'));

  const digest = sha256(body);
  if (digest !== meta.sha256) {
    problems.push(`the .jsonl hashes to ${digest}, the meta says ${meta.sha256}`);
  }

  const text = body.toString('utf8');
  const vectors = text.split('\n').filter((line) => line).map((line) => JSON.parse(line));
  if (STRAY_LINE_BREAK.test(text)) {
    problems.push('a line contains a character some line readers treat as a break');
  }
  if (vectors.length !== meta.count) {
    problems.push(`${vectors.length} lines, the meta says ${meta.count}`);
  }

  const assetText = readFileSync(join(directory, 'asset.json'), 'utf8');
  const assetSha = sha256(Buffer.from(assetText, 'utf8'));
  if (assetSha !== meta.asset_sha256) {
    problems.push(`the asset beside the vectors hashes to ${assetSha}, `
                  + `the meta says ${meta.asset_sha256}`);
  }
  const tokenizer = new Tokenizer(JSON.parse(assetText));

  let mismatched = 0;
  for (const v of vectors) {
    if (!sameArray(tokenizer.tokenize(v.input), v.tokens)
        || !sameArray(tokenizer.run(v.input), v.subtokens)) {
      mismatched += 1;
      if (mismatched <= 5) {
        problems.push(`${v.id}: ${JSON.stringify(v.input)} -> `
          + `${JSON.stringify(tokenizer.tokenize(v.input))} / `
          + `${JSON.stringify(tokenizer.run(v.input))}, expected `
          + `${JSON.stringify(v.tokens)} / ${JSON.stringify(v.subtokens)}`);
      }
    }
  }
  if (mismatched > 5) problems.push(`... and ${mismatched - 5} more mismatches`);

  const history = parseHistory(readFileSync(join(directory, 'test_vectors.history'), 'utf8'));
  if (!history.length) {
    problems.push('the history is empty');
  } else {
    const last = history[history.length - 1];
    if (last.count !== meta.count || last.sha256 !== meta.sha256) {
      problems.push(`the newest history line does not describe this file: ${last.line}`);
    }
    if (history.length >= 2) {
      const prev = history[history.length - 2];
      if (last.sha256 !== prev.sha256 && last.count === prev.count && !last.ruleChange) {
        problems.push('the digest changed while the count did not, and the history '
                      + `line does not say a rule changed: ${last.line}`);
      }
    }
  }

  const present = new Set(vectors.map((v) => v.feature));
  for (const feature of REQUIRED_FEATURES) {
    if (!present.has(feature)) problems.push(`no vector covers the feature '${feature}'`);
  }

  for (const p of problems) console.log(`FAIL  ${p}`);
  console.log(`${vectors.length} vectors, ${present.size} features, ${problems.length} problems`);
  return problems.length ? 1 : 0;
}

if (process.argv[1] === fileURLToPath(import.meta.url)) {
  process.exit(main(process.argv[2]));
}

File: site/tokenizer/check_test_vectors.py
# -*- coding: utf-8 -*-
"""§16.6's gate, the half of it that runs in Python. The JavaScript port runs the
same file and must reach the same verdict.

Four assertions, and the third is the one that exists because of how this can be
cheated:

1. The `.jsonl`'s bytes hash to what the meta says, and there are as many lines as it
   says. A vector file that has drifted from its own header is not evidence.
2. The Python implementation reproduces every triple.
3. **The digest may not change while the count stays the same** unless the history's
   newest line says a rule changed and why. Changing a tokenizer rule alters the
   expected output of thousands of existing vectors and leaves the count where it was
   — which is exactly the shape of a file quietly regenerated to match a broken
   implementation. A real rule change adds the cases the new rule needs (§16.5), so it
   moves the count anyway; if it genuinely does not, say so in the history line.
4. Every feature §16.5 names by hand is present. Sampling real expressions cannot
   reach pipeline steps 1, 2 and 3 at all, so a sample size is not coverage.
"""
import hashlib
import json
import os
import sys

_HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, os.path.normpath(os.path.join(_HERE, '..', '..', '..')))

# The features §16.5 requires by name. A vector file missing any of them is not a
# gate, whatever its size.
REQUIRED_FEATURES = (
    '16.2', '5.3', 'nfd', 'u007f', 'subsup_without_a_fold_entry', 'separator_only',
    'boundary_character', 'fallback_clause', 'private_use_escape',
    'escape_against_ascii_symbolic', 'astral_symbol', 'adjacent_fold_markers',
    'escape_scanning', 'numeric_class', 'empty', 'real_expression', 'real_name',
)


def parse_history(text):
    out = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith('#'):
            continue
        fields = dict(part.split('=', 1) for part in line.split() if '=' in part)
        out.append({'count': int(fields['count']), 'sha256': fields['sha256'],
                    'rule_change': 'rule-change:' in line, 'line': line})
    return out


def main(directory=_HERE, tokenizer=None):
    meta = json.load(open(os.path.join(directory, 'test_vectors.meta.json'), encoding='utf-8'))
    body = open(os.path.join(directory, 'test_vectors.jsonl'), 'rb').read()
    problems = []

    digest = hashlib.sha256(body).hexdigest()
    if digest != meta['sha256']:
        problems.append('the .jsonl hashes to %s, the meta says %s' % (digest, meta['sha256']))

    # Split on LF and nothing else: str.splitlines also breaks at U+0085, U+2028 and
    # U+2029, which real Isabelle text contains. The generator escapes those three for
    # the same reason, so a file that still splits differently under the two readings
    # is a malformed file and this catches it.
    text = body.decode('utf-8')
    vectors = [json.loads(line) for line in text.split('\n') if line]
    if len(text.split('\n')) - 1 != len(text.splitlines()):
        problems.append('a line contains a character some line readers treat as a break')
    if len(vectors) != meta['count']:
        problems.append('%d lines, the meta says %d' % (len(vectors), meta['count']))

    if tokenizer is None:
        from Isabelle_Semantic_Embedding.isabelle_tokenizer import Tokenizer
        from Isabelle_Semantic_Embedding import tokenizer_asset
        tokenizer = Tokenizer(tokenizer_asset.build_asset())
    mismatched = 0
    for v in vectors:
        if tokenizer.tokenize(v['input']) != v['tokens'] or tokenizer(v['input']) != v['subtokens']:
            mismatched += 1
            if mismatched <= 5:
                problems.append('%s: %r -> %s / %s, expected %s / %s'
                                % (v['id'], v['input'], tokenizer.tokenize(v['input']),
                                   tokenizer(v['input']), v['tokens'], v['subtokens']))
    if mismatched > 5:
        problems.append('... and %d more mismatches' % (mismatched - 5))

    history = parse_history(open(os.path.join(directory, 'test_vectors.history'),
                                 encoding='utf-8').read())
    if not history:
        problems.append('the history is empty')
    else:
        last = history[-1]
        if (last['count'], last['sha256']) != (meta['count'], meta['sha256']):
            problems.append('the newest history line does not describe this file: %s'
                            % last['line'])
        if len(history) >= 2:
            prev = history[-2]
            if last['sha256'] != prev['sha256'] and last['count'] == prev['count'] \
                    and not last['rule_change']:
                problems.append(
                    'the digest changed while the count did not, and the history line '
                    'does not say a rule changed: %s' % last['line'])

    present = {v['feature'] for v in vectors}
    for feature in REQUIRED_FEATURES:
        if feature not in present:
            problems.append('no vector covers the feature %r' % feature)

    for p in problems:
        print('FAIL  %s' % p)
    print('%d vectors, %d features, %d problems'
          % (len(vectors), len(present), len(problems)))
    return 1 if problems else 0


if __name__ == '__main__':
    sys.exit(main(*sys.argv[1:]))
